Add cameras

KevinBasener opened the pull request:
* Player camera
* MainCamera, that also is able to zoom in and out
* Switch between the two cameras
* Add signals for communication between player, main and map

KevinBasener committed:
Added all signals used to communicate between map player and main

jujujuok committed:
Add Player and Map Camera

jujujuok committed:
Camera follows MouseClick

jujujuok committed:
Update MainCamera.gd

KevinBasener committed:
Improved MainCamera

*Add drag movement via mouse wheel
*Add zoom function

KevinBasener committed:
Remove unnecessary print statements

hfxbse reviewed (changes requested):
Please for the love of god rebase your changes on to the `main`-branch to contain only the stuff that's relevant for the camera. I won't look at 46 files changed with 1615 lines added.

KevinBasener commented:
> Please for the love of god rebase your changes on to the `main`-branch to contain only the stuff that's relevant for the camera. I won't look at 46 files changed with 1615 lines added.

No its only 7 files

hfxbse commented:
Is there a possibility to set the "zoom"-level consistently? It still looks way to small on a high DPI screen and I think this is something that could be fixed with the camera.

KevinBasener committed:
Fix camera and sart implementation of the mobile camera

ubeF commented:
There seems to be no button for unlocking and moving the camera?

KevinBasener committed:
Add camera switch button

KevinBasener committed:
Add mobile hud

KevinBasener committed:
Change camera toggle button

KevinBasener committed:
Fixed Camera Button

KevinBasener committed:
Improve drag movement

KevinBasener commented:
All issues should be addressed

hfxbse committed:
pick right scaling mode

scaling the rendering canvas instead of stretching it later, which
uses the full screen and does not cause upscaling artifacts.

hfxbse commented on main/MainCamera.gd:
File name should also be in snake case as the rest.

hfxbse commented on main/MainCamera.gd:
```diff
@@ -0,0 +1,50 @@
+	if is_current():  
+		if event is InputEventMouseMotion:
+			if event.button_mask == MOUSE_BUTTON_MASK_LEFT:
+				position -= event.relative
```
The applied position offset should scale with the zoom level as it feels like you drag the camera over a sticky surface.

hfxbse replied:
Had a second look, apparently it already does it, but it still is not enough.

hfxbse commented on main/main.gd:
```diff
@@ -3,20 +3,41 @@ extends Node
 	print(question_handler.get_next_question(Question.Difficulty.EASY).question)
 	print(question_handler.get_random_question(Question.Difficulty.EASY).question)
-
+	
```
There is an unnecessary tab.

hfxbse reviewed (changes requested):
Not sure who it is done but when switching to the free camera it always jumps to the game origin which is a bit weird. Imo it should start a little bit zoomed out at the current player position.

I also set the correct scaling mode to not introduce up-scaling artifacts and use the whole screen aspect ratio with 50392e8f62f7ca024ba09dd2ab3c37d74618c62c. You need to anchor the camera button to the left bottom then to be in the correct position regardless of aspect ratio.

I also don't see a real reason to not place the HUD scene in the main scene, as it is always drawn. You can still remove the additional buttons if needed by referencing the HUD element in the scene. It requires less code an is less unexpected where the buttons will be placed.

As I would classify the button as an overlay, it might be appropriate to move stuff related to it in an directory called `overlay`.

hfxbse commented on main/MainCamera.gd:
```diff
@@ -0,0 +1,50 @@
+	zoom = lerp(zoom, 
+	target_zoom * Vector2.ONE,
+	zoom_rate * delta
+	)
```
The call to `lerp` should be on one line, or at least have correct indents.

hfxbse reviewed (changes requested):
White spaces are free :)

KevinBasener committed:
Fix camera toggle signals

KevinBasener committed:
Merge branch 'player' of https://github.com/hfxbse/mparty into player

KevinBasener committed:
Fix requested

KevinBasener committed:
Creat overlay directory

KevinBasener commented:
> Not sure who it is done but when switching to the free camera it always jumps to the game origin which is a bit weird. Imo it should start a little bit zoomed out at the current player position.
> 
> I also set the correct scaling mode to not introduce up-scaling artifacts and use the whole screen aspect ratio with [50392e8](https://github.com/hfxbse/mparty/commit/50392e8f62f7ca024ba09dd2ab3c37d74618c62c). You need to anchor the camera button to the left bottom then to be in the correct position regardless of aspect ratio.
> 
> I also don't see a real reason to not place the HUD scene in the main scene, as it is always drawn. You can still remove the additional buttons if needed by referencing the HUD element in the scene. It requires less code an is less unexpected where the buttons will be placed.
> 
> As I would classify the button as an overlay, it might be appropriate to move stuff related to it in an directory called `overlay`.

Should be addressed

hfxbse commented on main/main_camera.gd:
```diff
@@ -14,7 +14,7 @@ func _unhandled_input(event: InputEvent) -> void:
 		if event is InputEventMouseMotion:
 			if event.button_mask == MOUSE_BUTTON_MASK_LEFT:
-				position -= event.relative
+				position -= event.relative * zoom
```
When zooming, in the zoom value increases and when zooming out, the zoom value decreases.
It should be `position -= event.relative / zoom`. Multiplying made it even worse.

KevinBasener replied:
Good call

hfxbse commented on main/main_camera.gd:
```diff
@@ -31,20 +31,11 @@ func zoom_out():
+	zoom = lerp(zoom, target_zoom * Vector2.ONE, zoom_rate * delta)
 	set_physics_process(
 		not is_equal_approx(zoom.x, target_zoom)
 	)
```
`set_physics_process` might be on one line as well.

KevinBasener replied:
changed it

hfxbse reviewed (changes requested):
b086db4c462f1554642e8958c3239e55e43f2cf6 broke the main scene dependency on the HUD scene somehow and it will not let me select the correct dependency for an reason unknown to me.

250669ec616feffbe2ca89df885e8ef2c734e064 looked good besides white spaces so I think it is nearly ready to merge.

KevinBasener committed:
Fix requested changes hopefully final

KevinBasener commented:
> [b086db4](https://github.com/hfxbse/mparty/commit/b086db4c462f1554642e8958c3239e55e43f2cf6) broke the main scene dependency on the HUD scene somehow and it will not let me select the correct dependency for an reason unknown to me.
> 
> [250669e](https://github.com/hfxbse/mparty/commit/250669ec616feffbe2ca89df885e8ef2c734e064) looked good besides white spaces so I think it is nearly ready to merge.

Hope dependency doesn't break anymore

hfxbse committed:
adjust code formating
